tempesta_fw/tls.c

```diff
@@ -530,11 +530,12 @@ tfw_tls_send(TlsCtx *tls, struct sg_table *sgt, bool close)
 			       skb, skb->len);
 		}
 	}
-
-	if (close)
+	if (close) {
 		flags |= SS_F_CONN_CLOSE;
-	if (ttls_xfrm_ready(tls))
-		flags |= SS_F_ENCRYPT;
+		TFW_CONN_TYPE(&conn->cli_conn) |= Conn_Stop;
+	}
+	if (ttls_xfrm_ready(tls) && !was_encrypt)
```

Comment:
Having both `ttls_xfrm_ready()` and `was_encrypt` looks redundant and wrong.

We already know which records we have to encrypt and which not. In the `T_FSM_STATE(TTLS_SERVER_FINISHED)` we change `tls->state` before `__ttls_send_record()` is called, so here `tls->state` is a bit off. But order can be changed. Here is what I'm talking about:
```patch
diff --git a/tls/tls_srv.c b/tls/tls_srv.c
index 58699482..b9e2cb7f 100644
--- a/tls/tls_srv.c
+++ b/tls/tls_srv.c
@@ -2262,19 +2262,20 @@ ttls_handshake_finished(TlsCtx *tls)
 	}
 	T_FSM_STATE(TTLS_SERVER_FINISHED) {
 		if ((r = ttls_write_finished(tls, &sgt, &p)))
 			return r;
 		CHECK_STATE(TLS_HEADER_SIZE + TTLS_HS_FINISHED_BODY_LEN);
+		sg_mark_end(&sgt.sgl[sgt.nents - 1]);
+		r = __ttls_send_record(tls, &sgt, false, false);
 		/*
 		 * In case of session resuming, invert the client and server
 		 * ChangeCipherSpec messages order.
 		 */
 		tls->state = tls->hs->resume
 			     ? TTLS_CLIENT_CHANGE_CIPHER_SPEC
 			     : TTLS_HANDSHAKE_WRAPUP;
-		sg_mark_end(&sgt.sgl[sgt.nents - 1]);
-		return __ttls_send_record(tls, &sgt, false, true);
+		return r;
 	}
 	}
 	T_FSM_FINISH(r, tls->state);
 
 	/* If we exit here, then something went wrong. */
diff --git a/tls/ttls.c b/tls/ttls.c
index 5041ce56..1b0c6e89 100644
--- a/tls/ttls.c
+++ b/tls/ttls.c
@@ -242,11 +242,12 @@ EXPORT_SYMBOL(ttls_register_bio);
  */
 bool
 ttls_xfrm_ready(TlsCtx *tls)
 {
 	return tls->state >= TTLS_CLIENT_FINISHED
-	       && tls->state != TTLS_SERVER_CHANGE_CIPHER_SPEC;
+	       && tls->state != TTLS_SERVER_CHANGE_CIPHER_SPEC
+	       && tls->state != TTLS_SERVER_FINISHED;
 }
 EXPORT_SYMBOL(ttls_xfrm_ready);
 
 #if defined(TTLS_CLI_C)
 static int
```

I think, it should work. And then `was_encrypt` can be removed altogether.

Reply:
It will fall here
https://github.com/tempesta-tech/tempesta/blob/master/tls/ttls.c#L796

Reply:
Well, maybe. But anyway, `tls->state` should be enough to figure out if a record needs to be encrypted. Perhaps we'll need a new predicate for that.